### DISCUSSION TITLE
Support X-Forwarded-Prefix header

### DISCUSSION THREAD

**Discussion by StijnCaerts** [Kludex/uvicorn#1386] [Ideas] [answered]
Hi,

What are your thoughts on supporting the `X-Forwarded-Prefix` header used by proxies to tell an application it is served on a prefixed path. Currently, the `X-Forwarded-For` and `X-Forwarded-Proto` headers are supported in the `ProxyHeadersMiddleware`, so it would be nice if the `X-Forwarded-Prefix` header could be added as well.

**Comment by lovelydinosaur** [comment, marked as answer]
For use-cases such as this I'd suggest implementing the behaviour as ASGI middleware inside your own codebase.

**Reply by jankatins** [reply]
Is there a specific reason why this doesn't fit into the current ProxyHeadersMiddleware class? If it would be ok, I would be able to send a PR implementing this.

**Comment by StijnCaerts** [comment]
Ok, that works for me. Thank you for your opinion on this.

**Comment by jankatins** [comment]
I would have also expected that the `ProxyHeadersMiddleware` solves this and was especially surprised that fastapi doens't discuss an easy middleware solutions for this but points to some gunicorn cli argument :-(

This page has two middleware solutions for this problem: https://ecm.community/t/fast-api-und-x-forwarded-prefix/192 by @uwohlfeil

Copying the middleware class solution here:

```python
from typing import Union, cast
from starlette.types import ASGIApp, Scope, Receive, Send
from asgiref.typing import (
    HTTPScope,
    WebSocketScope,
)

class XPrefixHeaderFixMiddleWare:
    def __init__(self, app: ASGIApp):
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] in ("http", "websocket"):
            scope = cast(Union[HTTPScope, WebSocketScope], scope)
            headers = dict(scope["headers"])
            if b"x-forwarded-prefix" in headers:
                prefix = headers[b"x-forwarded-prefix"].decode("UTF-8")
                scope['root_path'] = prefix

        return await self.app(scope, receive, send)
```
